src/Data/Store/Streaming.hs

```diff
@@ -86,91 +101,103 @@ encodeMessage (Message x) =
 {-# INLINE encodeMessage #-}
 
 -- | The result of peeking at the next message can either be a
--- successfully deserialised 'Message', or a request for more input.
-data PeekMessage m a = Done (Message a)
-                     | NeedMoreInput (ByteString -> m (PeekMessage m a))
-
--- | Decode a 'Message' of known size from a 'ByteBuffer'.
-peekSized :: (MonadIO m, Store a) => ByteBuffer -> Int -> m (PeekMessage m a)
-peekSized bb n =
-    BB.unsafeConsume bb n >>= \case
-        Right ptr -> liftM (Done . Message) $ decodeFromPtr ptr n
-        Left _ -> return $ NeedMoreInput (\ bs -> BB.copyByteString bb bs
-                                                  >> peekSized bb n)
+-- successfully deserialised object, or a request for more input.
+type PeekMessage i m a = FT ((->) i) m a
```

Comment:
It wouldn't hurt to add a comment here, explaining how `FT` is what we want. The new `PeekMessage` is much more elegant and versatile, but a bit less obvious.